### ISSUE TITLE
How to provide our own daaset.json or csv file to the llama2:7B finetune using tune run

### ISSUE THREAD

**Issue by himanshushukla12** [pytorch/torchtune#1099]
I was trying to fine-tune using the `torchrun` technique and command line code looks like:
`tune run lora_finetune_single_device --config llama2/7B_lora_single_device batch_size=16 enable_activation_checkpointing=True max_steps_per_epoch=4` , can anyone help me to pass my json file as a dataset here.

**Comment by lulmer** [comment]
I have the same problem, I am missing something from the [documentation](https://pytorch.org/torchtune/main/tutorials/datasets.html). It would be nice to have more explanations/examples on this.

**Comment by RdoubleA** [comment]
@himanshushukla12 Thanks for the question, what type of dataset are you trying to train on? Do you require any formatting?

You can typically insert a custom dataset file using `torchtune.datasets.instruct_dataset`, `torchtune.datasets.chat_dataset`, or `torchtune.datasets.text_completion_dataset`. 
- Instruct data can map fields from your json quickly into a custom prompt template or one of the templates from the library (see [here](https://pytorch.org/torchtune/main/tutorials/datasets.html#custom-instruct-dataset-and-instruct-templates))
- Chat data is specifically for conversations, e.g., between user and AI assistant (see [here](https://pytorch.org/torchtune/main/tutorials/datasets.html#custom-chat-dataset-and-chat-formats) as an example). You should format your json like this [Message](https://github.com/pytorch/torchtune/blob/ef6e196d8e47e9bc584bc9f7ce836f646443381f/torchtune/data/_types.py#L14) dataclass. Or you can use a [converter](https://github.com/pytorch/torchtune/blob/main/torchtune/data/_converters.py) or write your own to get it in the right format in code
- Text completion data is unstructured raw text, similar to pretraining data

For all of these datasets, you can update the config to use your custom json like so (see [here](https://pytorch.org/torchtune/main/tutorials/datasets.html#local-and-remote-datasets) for details):
```
dataset:
  _component_: torchtune.datasets.instruct_dataset
  source: json
  split: train
  template: import.path.to.CustomTemplate
  data_files: path/to/my/data.json
```

If you can provide more details on your use case and the type of data you are using, I'm happy to provide more specific guidance.

We are always open to feedback on our documentation, so if there's a specific pain point you are struggling with let me know and I can update the docs. If the above info and doc links are still not clear, please let me know. I can see that custom chat data is not well explained in the docs, so I'll take that as an action item to improve.

**Comment by himanshushukla12** [comment]
> @himanshushukla12 Thanks for the question, what type of dataset are you trying to train on? Do you require any formatting?
> 
> You can typically insert a custom dataset file using `torchtune.datasets.instruct_dataset`, `torchtune.datasets.chat_dataset`, or `torchtune.datasets.text_completion_dataset`.
> 
> * Instruct data can map fields from your json quickly into a custom prompt template or one of the templates from the library (see [here](https://pytorch.org/torchtune/main/tutorials/datasets.html#custom-instruct-dataset-and-instruct-templates))
> * Chat data is specifically for conversations, e.g., between user and AI assistant (see [here](https://pytorch.org/torchtune/main/tutorials/datasets.html#custom-chat-dataset-and-chat-formats) as an example). You should format your json like this [Message](https://github.com/pytorch/torchtune/blob/ef6e196d8e47e9bc584bc9f7ce836f646443381f/torchtune/data/_types.py#L14) dataclass. Or you can use a [converter](https://github.com/pytorch/torchtune/blob/main/torchtune/data/_converters.py) or write your own to get it in the right format in code
> * Text completion data is unstructured raw text, similar to pretraining data
> 
> For all of these datasets, you can update the config to use your custom json like so (see [here](https://pytorch.org/torchtune/main/tutorials/datasets.html#local-and-remote-datasets) for details):
> 
> ```
> dataset:
>   _component_: torchtune.datasets.instruct_dataset
>   source: json
>   split: train
>   template: import.path.to.CustomTemplate
>   data_files: path/to/my/data.json
> ```
> 
> If you can provide more details on your use case and the type of data you are using, I'm happy to provide more specific guidance.
> 
> We are always open to feedback on our documentation, so if there's a specific pain point you are struggling with let me know and I can update the docs. If the above info and doc links are still not clear, please let me know. I can see that custom chat data is not well explained in the docs, so I'll take that as an action item to improve.


My CSV/JSON dataset contains <prompt> and <java code>, how to provide it model to train using torchtune

**Comment by himanshushukla12** [comment]
looks like `tune run lora_finetune_single_device  --config llama2/7B_lora_single_device dataset=torchtune.datasets.text_completion_dataset dataset.source="dataset.json" dataset.split=train dataset.column="prompt"` is the way to do this but somewhere in between I was not able to fix something.

**Comment by lulmer** [comment]
@joecummings What is unclear to me is : 
*  How torchtune is supposed to guess the json file dataset path here ? 
*  Same goes for the output/response field that is never mapped anywhere, how do we know torchtune will guess which field is it ?
*  In your message why is it `column_mapping` and not `column_map` like the signature of `AlpacaInstructTemplate` class is suggesting.
*  When writing a new template, where should it be saved exactly ? `import.path.to.CustomTemplate` is not explicit.

I think those points should be addressed in the documentation.

**Comment by ebsmothers** [comment]
Hi @lulmer thanks for bringing up all these points. 

> How torchtune is supposed to guess the json file dataset path here since it is not given explicitly anywhere ?

In the example given by @joecummings `dataset.source=json` is referring to the path. You can see from the API docs [here](https://pytorch.org/torchtune/main/generated/torchtune.datasets.instruct_dataset.html#torchtune.datasets.instruct_dataset) that the `source` arg corresponds to the `path` arg in Hugging Face's `load_dataset` API, which will infer data files given a local directory ([ref](https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path).)

> Same goes for the output/response field that is never mapped anywhere, how do we know torchtune will guess which field is it ?

Good question. If the output field is not provided in `column_map` it is assumed to be `output`. This can be seen [here](https://github.com/pytorch/torchtune/blob/f200da58c8f5007b61266504204c61a171f6b3dd/torchtune/datasets/_instruct.py#L92-L96) in the code. I think our description of `column_map` in the API docs (e.g. the instruct_dataset one linked above) needs to be explicit about the default values. As is, the statement we currently have ("If None, assume these are identical") is not very useful.

> In your CLI example, why is it column_mapping and not column_map like the signature of AlpacaInstructTemplate class is suggesting.

Yeah I believe this is a typo, it should be column map as you've rightly pointed out.

> When writing a new template, where should it be saved exactly ? import.path.to.CustomTemplate as specified in the documentation is not explicit.

I think any importable path should be fine here (so anything in `sys.path`, for example). Under the hood torchtune is just using `importlib.import_module` (see [here](https://github.com/pytorch/torchtune/blob/f200da58c8f5007b61266504204c61a171f6b3dd/torchtune/config/_utils.py#L74)). Personally when I'm making changes like this I just define in the same file as other similar components since it's easiest.. you can just change the class name without changing the path at all (this does assume you have the repo cloned, or at least the source files easily editable). Anyways I do agree that we can make this clearer in the documentation.

> In your CLI example why is it dataset and not dataset._component_, this is not the same logic as the other parameters.

This I think we do address in the docs (see [here](https://pytorch.org/torchtune/main/deep_dives/configs.html#overriding-components)), but it's in the config deep-dive rather than the datasets tutorial. Personally I do think this is the right place for it (I also don't want to bloat up the datasets tutorial with too many unrelated concepts). But open to any suggestions you have on how we can make this a bit more discoverable.

**Comment by lulmer** [comment]
Hi @ebsmothers, thank you for the clarifications, most of them make sense and helped me understanding better. 

> > How torchtune is supposed to guess the json file dataset path here since it is not given explicitly anywhere ?
> 
> In the example given by @joecummings `dataset.source=json` is referring to the path. You can see from the API docs [here](https://pytorch.org/torchtune/main/generated/torchtune.datasets.instruct_dataset.html#torchtune.datasets.instruct_dataset) that the `source` arg corresponds to the `path` arg in Hugging Face's `load_dataset` API, which will infer data files given a local directory ([ref](https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path).)

I agree that now that you pointed this I think it is clearer, what I would have liked to see (see suggestion in bold font) in the docs is : 
"You can pass in a Hugging Face dataset path to the source parameter in any of our builders to specify which dataset on the hub to download **or use from a local directory path.**" in [this chapter](https://pytorch.org/torchtune/main/tutorials/datasets.html#hugging-face-datasets) of the documentation.

> > Same goes for the output/response field that is never mapped anywhere, how do we know torchtune will guess which field is it ?
> 
> Good question. If the output field is not provided in `column_map` it is assumed to be `output`. This can be seen [here](https://github.com/pytorch/torchtune/blob/f200da58c8f5007b61266504204c61a171f6b3dd/torchtune/datasets/_instruct.py#L92-L96) in the code. I think our description of `column_map` in the API docs (e.g. the instruct_dataset one linked above) needs to be explicit about the default values. As is, the statement we currently have ("If None, assume these are identical") is not very useful.

> > When writing a new template, where should it be saved exactly ? import.path.to.CustomTemplate as specified in the documentation is not explicit.
> 
> I think any importable path should be fine here (so anything in `sys.path`, for example). Under the hood torchtune is just using `importlib.import_module` (see [here](https://github.com/pytorch/torchtune/blob/f200da58c8f5007b61266504204c61a171f6b3dd/torchtune/config/_utils.py#L74)). Personally when I'm making changes like this I just define in the same file as other similar components since it's easiest.. you can just change the class name without changing the path at all (this does assume you have the repo cloned, or at least the source files easily editable). Anyways I do agree that we can make this clearer in the documentation.

That makes sense! I think the documentation would clearly benefit from the explanation you've just given here.

> > In your CLI example why is it dataset and not dataset._component_, this is not the same logic as the other parameters.
> 
> This I think we do address in the docs (see [here](https://pytorch.org/torchtune/main/deep_dives/configs.html#overriding-components)), but it's in the config deep-dive rather than the datasets tutorial. Personally I do think this is the right place for it (I also don't want to bloat up the datasets tutorial with too many unrelated concepts). But open to any suggestions you have on how we can make this a bit more discoverable.

I agree with the placement on the config deep dive chapter. I just think that not maintaining the same convention for all parameters is confusing if you don't read this chapter on the docs. If you want to stay with this philosophy maybe a small one-liner with a ref pointing toward this link you just mentioned would be good to have on the dataset tutorial. 

I am also not comfortable with the wording "component", is supposed it a class or a function or both ? 
Would be interesting to have a small tutorial/example on "how to create your own custom component" in the doc even if copying any existing component code is straightforward.

**Comment by ebsmothers** [comment]
Thanks @lulmer for all this valuable feedback. Let me work with @RdoubleA to address some of these points in our documentation. It may take a little bit as I'm juggling a couple other things right now, but please bear with us. (Alternatively if you're interested in contributing any of this we would gladly review a PR, otherwise I promise we will get to it soon!)

The point about overriding of components is especially interesting. We went back and forth on this one a bunch. We're kinda inspired by Hydra's instantiate API in case you're familiar with that. They use `_target_` instead of `_component_` but it's a similar principle -- either one is used to identify a class or function by some importable path, and calling `instantiate(cfg)` on something like 

```
_component_: path.to.my.class_or_function
arg1: a
arg2: 2
```

is equivalent to invoking `class_or_function(arg1='a', arg2=2)`. We chose to drop the `_component_` in the CLI overrides because it was more intuitive to us, e.g. `model=lora_llama3_8b model.lora_rank=16` makes sense because `lora_rank` is a property of `model`. But like I said, we are inspired a bit by Hydra, and Hydra actually allows recursive instantiation (i.e. nested configs), where you can do stuff like `model=lora_llama3_8b model.embeddings=some_embeddings_class model.embeddings.embedding_dim=2048`. You can see that in this case including`_component_` after each class starts to become messy. 

Anyways a bit of a digression but just some extra context for you. I agree with your point, we should be a lot more explicit about the fact that we handle `_component_` differently from other command-line overrides.

**Comment by samadroohi** [comment]
I am trying to use InstructTemplate to finetune llama2 on a prediction task using the following command: 

`tune run lora_finetune_single_device --config llama2/7B_lora_single_device dataset=torchtune.datasets.instruct_dataset dataset.source=json dataset.template=torchtune.data.InstructTemplate dataset.column_map='{"instruction": "prompt", "input": "prediction"}' dataset.data_files=dataset.json`

I am getting the following error:

`raise ValueError(f"Invalid {class_type} class: '{component_name}'") from None
ValueError: Invalid InstructTemplate class: 'torchtune.data.InstructTemplate'`

Can anyone help me to pass my dataset using InstructTemplate?

**Comment by RdoubleA** [comment]
Thanks for the discussion @ebsmothers and @lulmer. I've created an issue #1123 to track this.

@samadroohi it looks like you're pointing to the base InstructTemplate class here. What would you like your prompt template to look like? Are you trying to use the `AlpacaInstructTemplate`?

**Comment by samadroohi** [comment]
> Thanks for the discussion @ebsmothers and @lulmer. I've created an issue #1123 to track this.
> 
> @samadroohi it looks like you're pointing to the base InstructTemplate class here. What would you like your prompt template to look like? Are you trying to use the `AlpacaInstructTemplate`?

Thanks for your reply, @RdoubleA . Yes, I am using InstructTemplate. Indeed, I have unsuccessfully tried both InstructTemplate and AlpacaInstructTemplate to finetune llama2 on a classification problem using a json dataset with "prompt" and "prediction" keys. In both settings I got the  following error: 
`raise ValueError(f"Invalid {class_type} class: '{component_name}'") from None ValueError: Invalid InstructTemplate class: 'torchtune.data.InstructTemplate'

**Comment by artisanclouddev** [comment]
I really want to use torchtune, but why are the contents of this document so independent? Can't there be a through tutorial to introduce the fine-tuning of the Llama3 model? Especially I want to have my own data format.

I feel like I understand everything, but I just can't connect the dots

**Comment by RdoubleA** [comment]
Closing as addressed by #1134, please reopen if you're still running into issues fine-tuning on custom data @artisanclouddev @samadroohi
